[Ho Wen Zhong] iP

wz27 opened the pull request:
# Duke Task Manager
> "Simplicity is prerequisite for reliability." - Edsger W. Dijkstra ([view](https://en.wikipedia.org/wiki/Edsger_W._Dijkstra))

Duke Task Manager allows you to manage all your tasks in one place. Features include:
* command-line interface
* familiar commands
* import/export of tasks

To use Duke Task Manager,
1. download the JAR.
2. open the JAR.
3. add tasks, deadlines, or events. 😀 

Best of all, it's ~~only $1.99~~ **FREE**!

Things you can do in Duke Task Manager:

- [x] Manage tasks, deadlines, and events
- [ ] Reminders (coming soon)

This is the driver method. It is called in the `main` method.
```javascript
public class Duke {
    public void run() {
        ...
    }
}
```

Commit:
Add Gradle support

wz27 committed:
Level-1

wz27 committed:
Level-2

wz27 committed:
Level-3

wz27 committed:
Level-4

wz27 committed:
A-TextUiTesting

wz27 committed:
Level-5

wz27 committed:
Level-6

wz27 committed:
Refactored addToList() and exceptions

wz27 committed:
Add load taskList from file

wz27 committed:
Level-7

wz27 committed:
Sync branch-Level-8 to branch-Level-7

wz27 committed:
Level-8

wz27 committed:
Merge branch-Level-7 to master

wz27 committed:
Merge branch 'master' into branch-Level-8

wz27 committed:
Retain branch-Level-8's duke.txt.orig

wz27 committed:
Merge branch-Level-8 to master

wz27 committed:
A-MoreOOP

wz27 committed:
A-Packages

wz27 committed:
A-JUnit

wz27 committed:
A-Jar

wz27 committed:
A-JavaDoc

wz27 committed:
A-JavaDoc

wz27 committed:
A-JavaDoc

wz27 committed:
Merge branch 'A-JavaDoc'

wz27 committed:
A-CodingStandard

wz27 committed:
A-CodingStandard

wz27 committed:
Merge branch 'branch-A-CodingStandard'

wz27 committed:
Level-9

wz27 committed:
Merge branch 'branch-Level-9'

rootkie commented on src/main/java/command/AddCommand.java:
```diff
@@ -0,0 +1,35 @@
+package command;
+import task.*;
```
Imported classes should always be listed explicitly. [here](https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements). May be can set your IDE settings to not automatically convert to wildcard imports.
```suggestion
import task.Task;
```

rootkie commented on src/main/java/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,29 @@
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.ChronoUnit;
+import task.*;
+import duke.*;
```
Similar situation with previous wildcard imports.

rootkie commented on src/main/java/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,29 @@
+    private String desc;
+    private LocalDate by;
+
+    public AddDeadlineCommand(String desc, String by) {
```
LocalDate.parse can throw a DateTimeException, should either catch it here or make your function explicitly throw the exception.

```suggestion
    public AddDeadlineCommand(String desc, String by) throws DateTimeException {
```

rootkie commented on src/main/java/command/AddEventCommand.java:
```diff
@@ -0,0 +1,29 @@
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.ChronoUnit;
+import task.*;
+import duke.*;
```
Similar situation with previous wildcard imports.

rootkie commented on src/main/java/command/AddEventCommand.java:
```diff
@@ -0,0 +1,29 @@
+    private String desc;
+    private LocalDate at;
+
+    public AddEventCommand(String desc, String at) {
```
LocalDate.parse can throw a DateTimeException, should either catch it here or make your function explicitly throw the exception.

```suggestion
    public AddEventCommand(String desc, String at) throws DateTimeException {
```

rootkie commented on src/main/java/command/AddTodoCommand.java:
```diff
@@ -0,0 +1,24 @@
+package command;
+import task.*;
+import duke.*;
```
Similar situation with previous wildcard imports.

rootkie commented on src/main/java/command/Command.java:
```diff
@@ -0,0 +1,27 @@
+package command;
+import duke.*;
+import task.*;
```
Similar situation with previous wildcard imports.

rootkie commented on src/main/java/command/DeleteCommand.java:
```diff
@@ -0,0 +1,62 @@
+package command;
+import task.*;
+import duke.*;
```
Similar situation with previous wildcard imports.

rootkie commented on src/main/java/command/DoneCommand.java:
```diff
@@ -0,0 +1,58 @@
+package command;
+import task.*;
+import duke.*;
```
Similar situation with previous wildcard imports.

rootkie commented on src/main/java/command/Command.java:
```diff
@@ -0,0 +1,27 @@
+     *
+     * @return Boolean indicating if Command exits program.
+     */
+    public abstract boolean isExit();
```
Consider making this a concrete method instead of abstract since only 1 command will return true. So just override it for ExitCommand and you don't have to implement it in all the other commands.

huyuxin0429 replied:
Agreed, extension of command will be a polymorphic nightmare if this have to be repeated over and over again.

rootkie commented on src/main/java/command/ExitCommand.java:
```diff
@@ -0,0 +1,42 @@
+package command;
+import task.*;
+import duke.*;
```
Similar situation with previous wildcard imports.

rootkie commented on src/main/java/command/FindCommand.java:
```diff
@@ -0,0 +1,23 @@
+package command;
+import task.*;
+import duke.*;
```
Similar situation with previous wildcard imports.

rootkie commented on src/main/java/command/ListCommand.java:
```diff
@@ -0,0 +1,34 @@
+package command;
+import task.*;
+import duke.*;
```
Similar situation with previous wildcard imports.

rootkie commented on src/main/java/command/DeleteCommand.java:
```diff
@@ -0,0 +1,62 @@
+     */
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        Task task = tasks.get(deleteIndex);
+        tasks.delete(deleteIndex);
```
If the index is negative or too big, it will throw an IndexOutOfBoundsException. Should probably check for it.

rootkie commented on src/main/java/command/DoneCommand.java:
```diff
@@ -0,0 +1,58 @@
+     */
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        tasks.done(doneIndex);
+
+        Task task = tasks.get(doneIndex);
```
If the index is negative or too big, it will throw an IndexOutOfBoundsException. Should probably check for it.

rootkie commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,64 @@
+package duke;
+
+import command.*;
+import task.*;
+import duke_exception.*;
```
Similar situation with previous wildcard imports.

rootkie commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,150 @@
+package duke;
+
+import command.*;
+import duke_exception.*;
```
Similar situation with previous wildcard imports.

rootkie commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,150 @@
+            } else if (fullCommand.equals("list")) {
+                return new ListCommand();
+            } else if (fullCommand.startsWith("done")) {
+                return detectDone(fullCommand);
+            } else if (fullCommand.startsWith("todo")) {
+                return detectTodo(fullCommand);
+            } else if (fullCommand.startsWith("deadline")) {
+                return detectDeadline(fullCommand);
+            } else if (fullCommand.startsWith("event")) {
+                return detectEvent(fullCommand);
+            } else if (fullCommand.startsWith("delete")) {
+                return detectDelete(fullCommand);
+            } else if (fullCommand.startsWith("find")){
+                return detectFind(fullCommand);
```
Function naming starting with **detect** make it sound like it has a boolean return value rather than a Command. May be try something like createDoneCommand(fullCommand).

rootkie commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,150 @@
+        }
+    }
+
+    public static FindCommand detectFind(String str) throws DukeFindException {
```
Missing javadocs for public method

rootkie commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,122 @@
+import java.util.List;
+import java.util.Scanner;
+
+import task.*;
+import duke_exception.*;
```
Similar situation with previous wildcard imports.

rootkie commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,122 @@
+     * @return List of Tasks.
+     * @throws DukeException If there are file access errors.
+     */
+    public List<Task> load() throws DukeException {
```
Did not catch DateTimeException that can be thrown by LocalDate.parse()

```suggestion
    public List<Task> load() throws DukeException, DateTimeException {
```

rootkie commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,97 @@
+     *
+     * @param index The index of the Task done.
+     */
+    public void done(int index) {
+        taskList.get(index).setDone();
+    }
```
Can trigger IndexOutOfBoundException. Probably consider handling it here or change method definition to throw such exception.

rootkie commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,97 @@
+     *
+     * @param index The index of the Task to delete.
+     */
+    public void delete(int index) {
+        taskList.remove(index);
+    }
```
Can trigger IndexOutOfBoundException. Probably consider handling it here or change method definition to throw such exception.

huyuxin0429 commented on src/main/java/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        // tasks
+        Deadline newDeadline = new Deadline(desc, by);
+        tasks.add(newDeadline);
+        // ui
+        String response = respond(newDeadline, tasks.size());
+        ui.showResponse(response);
+        // storage
+        storage.save(tasks);
+    }
+}
```
Repeated code chunk, consider abstracting this portion out?

huyuxin0429 commented on src/main/java/command/ExitCommand.java:
```diff
@@ -0,0 +1,42 @@
+     * @return True.
+     */
+    @Override
+    public boolean isExit() {
+        return true;
+    }
```
Would an exit command be eaiser?

huyuxin0429 reviewed (commented):
Consider a major refactoring of your code

wz27 committed:
Merge branch 'add-gradle-support' of https://github.com/wz27/ip into branch-A-Gradle

wz27 committed:
Merge branch 'branch-A-Gradle'

wz27 committed:
Level-10

wz27 committed:
Level-10

wz27 committed:
Merge branch 'branch-Level-10'

FergusMok commented on src/main/java/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,31 @@
+
+    @Override
+    public String execute(TaskList tasks, Ui ui, Storage storage) {
+        // tasks
+        Deadline newDeadline = new Deadline(desc, by);
+        tasks.add(newDeadline);
+        // ui
+        String response = respond(newDeadline, tasks.size());
+        String result = ui.showResponse(response);
+        // storage
```
Would it better to remove the comments? Your code is written very clear, and it is easy to understand. This would adhere better to the [conventions for commenting](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#do-not-repeat-the-obvious).

FergusMok commented on src/main/java/command/AddEventCommand.java:
```diff
@@ -0,0 +1,31 @@
+
+    @Override
+    public String execute(TaskList tasks, Ui ui, Storage storage) {
+        // tasks
+        Event newEvent = new Event(desc, at);
+        tasks.add(newEvent);
+        // ui
+        String response = respond(newEvent, tasks.size());
+        String result = ui.showResponse(response);
+        // storage
```
Would it be better to remove these comments? Same reasoning as the AddDeadlineCommand comment.

FergusMok commented on src/main/java/command/ExitCommand.java:
```diff
@@ -0,0 +1,44 @@
+     * @return Exit message.
+     */
+    public String respond() {
+        String response = "Bye. Hope to see you again soon!";
```
Would it better to make this string response into a [named variable](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-magic-numbers)? This would make the code easier to understand too.

FergusMok commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,194 @@
+        scrollPane.setVvalue(1.0);
+        scrollPane.setFitToWidth(true);
+
+        // You will need to import `javafx.scene.layout.Region` for this.
```
Would it better to remove this comment? Your code is written clearly, and other users will most likely understand the reasoning for this.

FergusMok commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,150 @@
+     */
+    public static Command parse(String fullCommand) throws DukeException {
+        try {
+            if (fullCommand.equals("bye")) {
+                return new ExitCommand();
+            } else if (fullCommand.equals("list")) {
+                return new ListCommand();
+            } else if (fullCommand.startsWith("done")) {
+                return detectDone(fullCommand);
+            } else if (fullCommand.startsWith("todo")) {
+                return detectTodo(fullCommand);
+            } else if (fullCommand.startsWith("deadline")) {
+                return detectDeadline(fullCommand);
+            } else if (fullCommand.startsWith("event")) {
+                return detectEvent(fullCommand);
+            } else if (fullCommand.startsWith("delete")) {
+                return detectDelete(fullCommand);
+            } else if (fullCommand.startsWith("find")){
```
Would it be better to have these commands as named variables instead? It would also make the code clearer to other users.

FergusMok reviewed (commented):
Hello, your code is well-written and very clear, especially for the Exceptions. Just a few small suggestions to adhere better to the conventions.

wz27 committed:
Update build.gradle file

application.mainClassName variable: point to Launcher.java

The build.gradle file needs to point to the correct main class to start
the Duke program.

Changing the main class from Duke.java to Launcher.java allows the JAR
file to execute properly.

Let's change the mainClassName variable to Launcher.java.

wz27 committed:
Update to follow coding standards

wz27 committed:
A-Assertions

wz27 committed:
Update code to improve code quality

There are some areas where code quality can be improved upon.

Production code needs to be of high quality so that it is readable and
understandable.

Let's improve the code quality for every Java file in the project.

wz27 committed:
Add Java assertions

There are no assertions used in the program.

Assertions document important assumptions that should hold at various
points in the code so that the runtime can verify them.

Add Java assertions at important points in the code, such as during
initialization.

An assertion failure indicates a possible bug in the code. It alerts us
to a possible bug at important points and stops execution to be safe.

wz27 committed:
Merge pull request #1 from wz27/branch-A-Assertions

PR branch-A-Assertions

wz27 committed:
Merge branch 'master' into branch-A-CodeQuality

wz27 committed:
Merge pull request #2 from wz27/branch-A-CodeQuality

Update code to improve code quality

wz27 committed:
Add gradle.yml file to set up Continuous Integration

There is no Continuous Integration for this project.

GitHub Actions can run workflow specified in .yml file when project
events are triggered, such as build tests and IO tests.

Let's add gradle.yml file to set up Continuous Integration.

wz27 committed:
Add C-Undo extension

The program does not support undo function.

Undo allows user to undo the most recent command.

Let's add undo functionality to Duke.

wz27 committed:
Merge branch 'branch-C-Undo'

wz27 committed:
A-UserGuide

wz27 committed:
Update README.md

wz27 committed:
A-BetterGui
